app/server/api/token_api.py

```diff
@@ -61,8 +62,62 @@ def post(self):
 
         return make_response(jsonify(response_object)), 201
 
+
+class TokenExchangeContractAPI(MethodView):
+
+    @requires_auth(allowed_basic_auth_types=['internal'])
+    def put(self, token_id):
+
+        put_data = request.get_json()
+        exchange_contract_id = put_data.get('exchange_contract_id')
+
+        # Even though it's non-standard, allow specifying exchange contract by address
+        # Because it's often way easier to get hold of than the contract id
+        exchange_contract_address = put_data.get('exchange_contract_address')
+
+
+        token = Token.query.get(token_id)
+
+        if token is None:
+            return make_response(jsonify(
+                {'message': 'No Token found for ID: {}'.format(token_id)}))
+
+
+        if exchange_contract_address and exchange_contract_id:
+            return make_response(jsonify(
+                {'message': 'Must ONLY specify one of exchange contract id or address'}))
+
+        if exchange_contract_id:
+            exchange_contract = ExchangeContract.query.get(exchange_contract_id)
+
+        elif exchange_contract_address:
+            exchange_contract = ExchangeContract.query.filter_by(blockchain_address=exchange_contract_address).first()
+
+        else:
+            return make_response(jsonify(
+                {'message': 'Must specify one of exchange contract id or address'}))
+
+        if exchange_contract is None:
+            return make_response(jsonify(
+                {'message': 'No Exchange Contract found for ID: {}'.format(exchange_contract_id)}))
+
+        token.add_exchange_contract(exchange_contract)
```

Comment:
hmm so a token has an exchange contract and exchange contract has a `reserve_token_id`. is that the token you exchange to, and the contract defines the rate?

Reply:
Close. Even if you're exchanging between to CIC tokens (aka ones defined by a bonding curve WRT a reserve), you still go through the reserve token to make the full exchange. so you could have an exchange that is `CIC1 > RESERVE` but you could also have one that is `CIC1 > RESERVE > CIC2`